Introduction
============

.. raw:: html

   <div style="display: flex; gap: 20px; flex-wrap: wrap;">

     <figure style="flex: 1; text-align: center;">
       <img src="_static/kcl_logo.png" style="width:70%;">
     </figure>

     <figure style="flex: 1; text-align: center;">
       <img src="_static/sbf_logo.png" style="width:100%;">
     </figure>

   </div>

At King's College London, the `Spatial Biology Facility <https://www.kcl.ac.uk/research/facilities/the-spatial-biology-facility>`_ utilises both CosMx and Xenium technologies. We have identified a gap in the integration of spatial transcriptomics data, from raw output to Python-compatible formats, regardless of whether the data originates from 10X Xenium or NanoString CosMx instruments.

To address this, we have developed MOSAIK (*Multi-Origin Spatial Transcriptomics Analysis and Integration Kit*): a dynamic, evolving workflow designed to fully harness the power of spatial biology and support seamless multimodal integration.

MOSAIK is built to adapt to the latest versions of key platform providers, including 10X Genomics and NanoString, while staying aligned with emerging trends in the spatial biology community.


